kernel-plugins/src/BLSValidator.sol

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.8.4 <0.9.0;
+
+import "kernel/src/validator/IValidator.sol";
+import "openzeppelin-contracts/contracts/utils/cryptography/EIP712.sol";
+import "kernel/src/utils/KernelHelper.sol";
+import "account-abstraction/samples/bls/lib/hubble-contracts/contracts/libs/BLS.sol";
+
+// BLSValidator is a validator that uses BLS signatures to validate transactions.
+// TODO: Consider account recovery, aggregate signatures, and use EIP 712.
+// Note: This is not audited and should not be used in production.
+contract BLSValidator is IKernelValidator {
+    bytes32 public constant BLS_DOMAIN = keccak256("eip4337.bls.domain");
+    
+    event OwnerChanged(address indexed kernel, uint256[4] indexed oldPublicKey, uint256[4] indexed newPublicKey);
+
+    mapping(address => uint256[4]) public blsValidatorStorage;
+
+    function disable(bytes calldata) external override {
+        delete blsValidatorStorage[msg.sender];
+    }
+
+    function enable(bytes calldata _data) external override {
+        require(_data.length >= 128, "Calldata is not long enough for bls public key");
+
+        uint256[4] memory publicKey = abi.decode(_data, (uint256[4]));
+        uint256[4] memory oldPublicKey = blsValidatorStorage[msg.sender];
+        blsValidatorStorage[msg.sender] = publicKey;
+
+        emit OwnerChanged(msg.sender, oldPublicKey, publicKey);
+    }
+
+    function validateUserOp(UserOperation calldata _userOp, bytes32 _userOpHash, uint256)
+        external
+        view
+        override
+        returns (uint256 validationData)
+    {
+        uint256[4] memory publicKey = blsValidatorStorage[_userOp.sender];
+        bytes memory hashBytes = abi.encodePacked(_userOpHash);
+
+        uint256[2] memory message = BLS.hashToPoint(
+            BLS_DOMAIN,
+            hashBytes
+        );
+        uint256[2] memory decodedSignature = abi.decode(_userOp.signature, (uint256[2]));
+        (bool verified, bool callSuccess) = BLS.verifySingle(decodedSignature, publicKey, message);
+
+        if (verified && callSuccess) {
+            return 0;
+        }
+        // TODO: check if wallet recovered
+        return SIG_VALIDATION_FAILED;
+    }
+
+
+    function validateSignature(bytes32 hash, bytes calldata signature) public view override returns (uint256) {
+        uint256[4] memory publicKey = blsValidatorStorage[msg.sender];
+        uint256[2] memory decodedSignature = abi.decode(signature, (uint256[2]));
```

Comment:
This might be overcomplicating things for an initial PR, but was just cross-referencing this with code in the v1 verification gateway. Do we want to add checks like the following at this stage?
```solidity
require(signature.length == 64, "Sig bytes length must be 64");
```

Or if we want to move to custom errors:
```solidity
if (signature.length != 64) revert SIG_BYTES_LENGTH_MUST_BE_64();
```

Same with checking `publicKey` is not `[0,0,0,0]`.

Again, might be overcomplicating things for now but thought I'd ask just in case.

Reply:
No I think that's a good call.  It took me so long just to get something working that I forgot about this.  I can probably go back now and add some validation/checks